pipeline/pipeline.go

```diff
@@ -132,312 +113,211 @@ func (p *Pipeline) isOutputModule(name string) bool {
+		return fmt.Errorf("invalid request: start block %d smaller that request outputs for module: %q start block %d", p.reqCtx.StartBlockNum(), module.Name, module.InitialBlock)
 	}
+	p.moduleHashes.HashModule(p.reqCtx.Request().Modules, module, p.graph)
+	return nil
 }
 
 func (p *Pipeline) Init(workerPool *orchestrator.WorkerPool) (err error) {
-	ctx := p.context
-	traceID := GetTraceID(ctx)
-	p.logger = p.logger.With(zap.Strings("outputs", p.request.OutputModules), zap.Bool("sub_request", p.isSubrequest), zap.String("trace_id", traceID.String()))
-
-	ctx, span := p.tracer.Start(ctx, "pipeline_init")
-	defer span.End()
+	p.reqCtx.StartSpan("pipeline_init")
 	defer func() {
 		if err != nil {
-			span.SetStatus(codes.Error, err.Error())
+			p.reqCtx.SetErr(err)
 		} else {
-			span.SetStatus(codes.Ok, "")
+			p.reqCtx.SetOk("")
 		}
+		p.reqCtx.EndSpan()
 	}()
 
-	p.logger.Info("initializing handler", zap.Uint64("requested_start_block", p.requestedStartBlockNum), zap.Uint64("requested_stop_block", p.request.StopBlockNum), zap.Bool("is_backprocessing", p.isSubrequest), zap.Strings("outputs", p.request.OutputModules))
-
-	p.moduleOutputCache = outputs.NewModuleOutputCache(p.outputCacheSaveBlockInterval, p.logger)
+	p.reqCtx.logger.Info("initializing handler",
+		zap.Uint64("requested_start_block", p.reqCtx.StartBlockNum()),
+		zap.Uint64("requested_stop_block", p.reqCtx.StopBlockNum()),
+		zap.Bool("is_back_processing", p.isSubrequest),
+		zap.Strings("outputs", p.reqCtx.Request().OutputModules),
+	)
 
-	if err := p.build(); err != nil {
-		span.SetStatus(codes.Error, err.Error())
-		return fmt.Errorf("building pipeline: %w", err)
+	if err := p.validateBinaries(); err != nil {
+		return fmt.Errorf("binary validation failed: %w", err)
 	}
 
-	p.computeModuleHashes()
-
-	for _, module := range p.modules {
-		isOutput := p.outputModuleMap[module.Name]
+	modules, storeModules, err := p.getModules()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve modules: %w", err)
+	}
 
-		if isOutput && p.requestedStartBlockNum < module.InitialBlock {
-			err := fmt.Errorf("invalid request: start block %d smaller that request outputs for module: %q start block %d", p.requestedStartBlockNum, module.Name, module.InitialBlock)
-			return err
+	moduleHashes := map[string]*pbsubstreams.Module{}
+	for _, module := range modules {
+		if err := p.validateAndHashModule(module); err != nil {
+			return fmt.Errorf("module failed validation: %w", err)
 		}
+		moduleHashes[p.moduleHashes.Get(module.Name)] = module
+	}
 
-		hash := p.moduleHashes.Get(module.Name)
-		_, err := p.moduleOutputCache.RegisterModule(module, hash, p.baseStateStore)
-		if err != nil {
-			return fmt.Errorf("registering output cache for module %q: %w", module.Name, err)
-		}
+	p.reqCtx.logger.Info("priming caching engine")
+	if err := p.cachingEngine.Prime(moduleHashes); err != nil {
+		return fmt.Errorf("failed to prime caching engine: %w", err)
```

Comment:
unsure about the rational to pry open the modules hashing thing, which was all containerized in that `ModulesHashes` object, especially when I see that the caching engine would benefit from that exact abstraction.